### PR TITLE
Initial attempt at Synapse strategy

### PR DESCRIPTION
**Pull request by numan**

### CODE REVIEW

File: contracts/Strategy.sol [hunk 1]
```diff
@@ -72,24 +143,79 @@ contract Strategy is BaseStrategy {
...
+
+            // If we still have LP tokens after trying to withdraw everything in `want`,
+            // withdraw in anything else that we can get
+            if (_lpTokenBalance > 0) {
+                uint256[] memory minAmounts = new uint256[](3);
+                //todo: should we try to not withdraw any amount in usdc?
+                minAmounts[0] = 0; // nUSD
+                minAmounts[1] = 0; // USDC
+                minAmounts[2] = 0; // fUSDT
+                syn3PoolSwap.removeLiquidity(
```
Comment (wavey0x) [inline contracts/Strategy.sol:196]: Since we want only USDC, should use this function instead.

```
function removeLiquidityOneToken(
        uint256 tokenAmount,
        uint8 tokenIndex,
        uint256 minAmount,
        uint256 deadline
    ) external returns (uint256);
```

File: contracts/Strategy.sol [hunk 2]
```diff
@@ -133,8 +259,158 @@ contract Strategy is BaseStrategy {
...
+            block.timestamp
+        );
+    }
+
+    function _unstakeLPTokens(uint256 _amount) internal {
+        synStakingMC.withdraw(pid, _amount, address(this));
+    }
+
+    function _withdrawLiquidity(uint256 _lpAmount) internal {
+        syn3PoolSwap.removeLiquidityOneToken(
```
Comment (wavey0x) [inline contracts/Strategy.sol:317]: Critical: 
in order to avoid unexpected losses on withdrawal, we should enforce a slippage tolerance using `minAmount` param in the `removeLiquidityOneToken` function.

To compute it, we should introduce a couple new storage values (with setters so that they can be adjusted) which will define the % loss we can tolerate:

1. `maxSlippageIn` (e.g. 0.05%)
2. `maxSlippageOut` (e.g. 0.05%)

Next, we can convert the `_lpAmount` to amount of want using `syn3PoolSwap.getVirtaulPrice()`.
We can then decrease this amount by the `maxSlippageOut` to find our slippage threshold. Now, any attempt to add/remove liquidity that yields less than we expect will revert.

File: contracts/Strategy.sol (hunk 1 above)
Comment (wavey0x) [reply contracts/Strategy.sol:196]: I would actually remove this custom logic from `liquidateAllPositions`.
To make things simpler, you can have this function call to `liquidatePosition`.
```
function liquidateAllPositions() internal override returns (uint256 liquidated) {
        liquidatePosition(stakedLPBalance());    // however, we will might want to create a separate function for this
        liquidated = balanceOfWant();
        return liquidated;
}
```

File: contracts/Strategy.sol [hunk 3]
```diff
@@ -72,24 +143,79 @@ contract Strategy is BaseStrategy {
...
+            _withdrawLiquidity(_lpTokenBalance);
+
+            _lpTokenBalance = unstakedLPBalance();
+
+            // If we still have LP tokens after trying to withdraw everything in `want`,
+            // withdraw in anything else that we can get
+            if (_lpTokenBalance > 0) {
+                uint256[] memory minAmounts = new uint256[](3);
+                //todo: should we try to not withdraw any amount in usdc?
+                minAmounts[0] = 0; // nUSD
```
Comment (wavey0x) [inline contracts/Strategy.sol:193]: remove this line.
0 is default vaule. so no need to set it. in fact, it actually consumes extra gas to re-assign value to 0

File: contracts/Strategy.sol [hunk 4]
```diff
@@ -72,24 +143,79 @@ contract Strategy is BaseStrategy {
...
+                minAmounts[1] = 0; // USDC
```
Comment (wavey0x) [inline contracts/Strategy.sol:194]: remove this line

File: contracts/Strategy.sol [hunk 5]
```diff
@@ -72,24 +143,79 @@ contract Strategy is BaseStrategy {
...
+                minAmounts[2] = 0; // fUSDT
```
Comment (wavey0x) [inline contracts/Strategy.sol:195]: remove this line

File: contracts/Strategy.sol [hunk 6]
```diff
@@ -72,24 +143,79 @@ contract Strategy is BaseStrategy {
...
-        uint256 totalAssets = want.balanceOf(address(this));
-        if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+        uint256 _liquidWant = wantBalance();
+
+        _amountNeeded = Math.min(_amountNeeded, estimatedTotalAssets()); // Otherwise we can end up declaring a liquidation loss when _amountNeeded is more than we own
+
+        if (_liquidWant < _amountNeeded) {
+            _unstakeLPTokens(stakedLPBalance());
```
Comment (wavey0x) [inline contracts/Strategy.sol:151]: why unstake the full amount when we can compute the proper amount and withdraw only what we need?

File: contracts/Strategy.sol [hunk 7]
```diff
@@ -10,38 +8,78 @@ import {
...
-        return "Strategy<ProtocolName><TokenType>";
+        return "StrategySynapseUSDC";
     }
 
     function estimatedTotalAssets() public view override returns (uint256) {
-        // TODO: Build a more accurate estimate using the value of all positions in terms of `want`
-        return want.balanceOf(address(this));
+        // Incase we aren't able to withdraw our deposit all in `want`, we might have other types of stable coin tokens
+        return
+            nusdBalance().add(usdtBalance()).add(wantBalance()).add(
```
Comment (wavey0x) [inline contracts/Strategy.sol:80]: i would remove all this and make estimatedTotalAssets the sum of
- want value of the staked LPs
- want value of unstaked LPs
- add want.balanceOf(strategy)

we typically dont use the farmed token value in estimatedTotalAssets.

File: contracts/Strategy.sol [hunk 8]
```diff
@@ -133,8 +259,158 @@ contract Strategy is BaseStrategy {
...
+            address(this),
+            block.timestamp
+        );
+    }
+
+    function _addliquidity(uint256 _amount) internal {
+        _checkAllowance(address(syn3PoolSwap), address(want), _amount);
+
+        uint256[] memory liquidityToAdd = new uint256[](3);
+        liquidityToAdd[0] = 0; // nUSD
```
Comment (wavey0x) [inline contracts/Strategy.sol:301]: remove this line

File: contracts/Strategy.sol [hunk 9]
```diff
@@ -133,8 +259,158 @@ contract Strategy is BaseStrategy {
...
+        liquidityToAdd[1] = _amount; // USDC
+        liquidityToAdd[2] = 0; // fUSDT
```
Comment (wavey0x) [inline contracts/Strategy.sol:303]: remove this line

**Review by wavey0x — COMMENTED** [review]
This strategy is a great guide for how you should approach accounting in a single sided strategy. It also uses masterchef for rewards, which makes it very similar to what you're trying to build.

https://github.com/tonkers-kuma/strategy-ssb/blob/beethoven/contracts/Strategy.sol

File: contracts/Strategy.sol (hunk 1 above)
Comment (numan) [reply contracts/Strategy.sol:196]: Since this is used in emergencies, does it make sense to call `liquidatePosition(stakedLPBalance());` and then call 

```
    function removeLiquidity(
        uint256 amount,
        uint256[] calldata minAmounts,
        uint256 deadline
    ) external returns (uint256[] memory);
```

Incase there isn't enough liquidity of `want`? Is that overkill?

Comment (numan) [reply contracts/Strategy.sol:196]: Actually, that's exactly what it is doing now:

1. `_withdrawLiquidity` calls `removeLiquidityOneToken`
2. If we still have an LP token balance after, it calls `removeLiquidity` to get anything else we can get our hands on

File: contracts/Strategy.sol (hunk 6 above)
Comment (numan) [reply contracts/Strategy.sol:151]: I guess I wasn't quite sure about how to deal with unexpected losses on withdraws. I'll try to think of a better solution
